atc/exec/task_step.go

```diff
@@ -156,8 +156,11 @@ func (step *TaskStep) run(ctx context.Context, state RunState) error {
 	var taskVars []vars.Variables
 
 	if step.plan.ConfigPath != "" {
-		// external task - construct a source which reads it from file
-		taskConfigSource = FileConfigSource{ConfigPath: step.plan.ConfigPath, Client: step.workerClient}
+		// external task - construct a source which reads it from file, and apply base resource type defaults.
+		taskConfigSource = BaseResourceTypeDefaultsApplySource{
+			ConfigSource:  FileConfigSource{ConfigPath: step.plan.ConfigPath, Client: step.workerClient},
+			ResourceTypes: step.plan.VersionedResourceTypes,
+		}
```

Comment:
I noticed in testing that this change does not apply to builds run with `fly execute`. In that case, the plan is already constructed (by fly, which doesn't know about the base resource type default settings), so we are in the `else` branch below. Locally, I changed line 177 below
```go
		taskConfigSource = StaticConfigSource{Config: step.plan.Config}
```
to
```go
		taskConfigSource = BaseResourceTypeDefaultsApplySource{
			ConfigSource: StaticConfigSource{Config: step.plan.Config},
			ResourceTypes: step.plan.VersionedResourceTypes,
		}
```
and this resulted in fly executed builds having the defaults applied, as expected. I did not check to see if this broke anything else, yet. 

I think it would be convenient to have parity between one-off builds and regular pipeline builds, as it may be confusing if they behave in subtly different ways - not limited to registry mirror settings. It's not about whether the volume of one-off builds potentially not using a configured mirror is problematic, but just the fact that different behavior is undesirable.

Reply:
Good catch! Your comment made me think over again. I think it makes sense to apply defaults after interpolate vars, because image_source itself could be defined by vars.